Use electron app

danigm opened the pull request:
https://phabricator.endlessm.com/T32216

danigm committed:
Use electron app for the frontend

https://phabricator.endlessm.com/T32216

dylanmccall commented on requirements-win.in:
```diff
@@ -1,3 +1,2 @@
 PyInstaller==3.6
 wxPython==4.1.1
```
Can we remove wxPython here as well? We aren't using any of pyeverywhere's runtime code.

danigm replied:
Yes, we can remove that!

dylanmccall commented on src/main.py:
```diff
@@ -154,128 +154,10 @@ def _(text):
 from kolibri_tools.utils import start_kolibri_server
 
 
-class MenuEventHandler:
```
Unused imports in main.py after these removals:

```
import json
import time
import webbrowser

from pew.ui import PEWShortcut
```

dylanmccall commented on kolibri-electron/src/index.js:
```diff
@@ -0,0 +1,90 @@
+
+const waitForKolibriUp = (splash, mainWindow) => {
+  console.log("Pinging kolibri server");
+  http.get(`${KOLIBRI}/static/assets/logo.ico`, (response) => {
```
A better URL to poll might be...

```suggestion
  http.get(`${KOLIBRI}/api/public/info`, (response) => {
```

Static files might actually be on a different origin (and served by a different application) in a future version of Kolibri - see learningequality/kolibri#7920. So, polling a static file might not tell us what we expect. And that API provides a slightly smaller response along with some potentially useful information :)

dylanmccall commented on kolibri-electron/src/index.js:
```diff
@@ -0,0 +1,90 @@
+  console.log("Pinging kolibri server");
+  http.get(`${KOLIBRI}/static/assets/logo.ico`, (response) => {
+    if (response.statusCode !== 200) {
+      console.log("Error: " + response.statusCode);
```
It is possible to show an error in the loading screen, as well. The Python app did it like this...

```
pew.ui.run_on_main_thread(self.view.evaluate_javascript, 'show_error()')
```

It can also show a message if loading is taking a long time:

```
pew.ui.run_on_main_thread(self.view.evaluate_javascript, 'show_retry()')
```

It would be a bit of a shame to lose that and looks like it can be done fairly nicely here. Something like `splash.webContents.executeJavascript("show_error()")` depending on a timestamp?

danigm replied:
You are right, I've updated the code to show the errors. I've even improved what we had before, reloading the kolibri backend if it's taking too long.

dylanmccall commented on kolibri-electron/src/index.js:
```diff
@@ -0,0 +1,90 @@
+  mainWindow.on('page-title-updated', (event) => {
+    event.preventDefault();
+  });
+
```
One small detail here: external links are opening in the application. Instead, they should open in the user's default browser. For that, we need to use either the `will-navigate` event or the `setWindowOpenHandler` method (maybe both?!) on `mainWindow.webContents` (https://www.electronjs.org/docs/api/web-contents).

Something like this, although better if it's rewritten slightly (import at the top with the others, etc).

```suggestion
  const { shell } = require('electron');

  const windowOpenHandler = ({url}) => {
    if (url.startsWith('file:') || url.startsWith(KOLIBRI)) {
      return {action: 'allow'};
    }
    shell.openExternal(url);
    return {action: 'deny'};
  };

  splash.webContents.setWindowOpenHandler(windowOpenHandler);
  mainWindow.webContents.setWindowOpenHandler(windowOpenHandler);
```

danigm replied:
Thanks a lot, I've modified a bit the suggested code to allow relative paths

dylanmccall reviewed (changes requested):
This looks reasonable to me. Some bits and pieces that could be cleaned up, but it works. Nice stuff, @danigm :)

danigm committed:
electron: Add development instructions

This patch adds a README file in the kolibri-electron with some simple
instructions to build and also adds the app icon.

https://phabricator.endlessm.com/T32216
